R/bs-theme-preset-brand.R

```diff
@@ -0,0 +1,830 @@
+  theme <- b_get(brand, "color")
+
+  for (field in names(b_get(brand, "color"))) {
+    if (field == "palette") {
+      theme[[field]] <- lapply(
+        rlang::set_names(names(theme[[field]])),
+        b_get_color,
+        brand = brand
+      )
+    } else {
+      theme[[field]] <- b_get_color(brand, field)
+    }
+  }
+
+  # Then replace brand.color with resolved colors
+  brand[["color"]] <- theme
+  brand
+}
+
+b_get_color <- function(brand, key) {
+  if (!b_has(brand, "color")) {
+    return(key)
+  }
+
+  theme_colors <- brand[["color"]]
+  theme_colors$palette <- NULL
+  palette <- brand[["color"]][["palette"]] %||% list()
+
+  key_og <- key
+  visited <- c()
+
+  cycle <- function(key) {
+    path <- c(visited, key)
+    if (length(path) > 10) {
+      path <- c(path[1:2], "...", path[-(1:(length(path) - 2))])
+    }
+    paste(path, collapse = " -> ")
+  }
+
+  assert_no_cycles <- function(key) {
+    if (key %in% visited) {
+      abort(
+        c(
+          sprintf(
+            "Cyclic references detected in `brand.color` for color '%s'.",
+            key_og
+          ),
+          "i" = cycle(key)
+        )
+      )
+    }
+    visited <<- c(visited, key)
+  }
+
+  p_key <- function(key) paste0("palette.", key)
+  value <- ""
+  i <- 0
+  while (value != key) {
+    i <- i + 1
+    if (i > 100) {
+      abort(
+        c(
+          sprintf(
+            "Max recursion limit reached while trying to resolve color '%s' using `brand.color`.",
+            key_og
+          ),
+          i = cycle(key)
+        )
+      )
+    }
+
+    in_theme <- key %in% names(theme_colors)
+    in_theme_unseen <- in_theme && !key %in% visited
+    in_pal <- key %in% names(palette)
+
+    if (in_pal && !in_theme_unseen) {
+      # Prioritize palette if theme was already visited
+      assert_no_cycles(p_key(key))
+      key <- palette[[key]]
+    } else if (in_theme) {
+      assert_no_cycles(key)
+      key <- theme_colors[[key]]
+    } else {
+      value <- key
+    }
+  }
+
+  return(value)
+}
+
+# Brand utilities --------------------------------------------------------------
+
+b_has <- function(brand, ...) {
+  x <- brand
+
+  for (f in c(...)) {
+    if (is.null(x[[f]])) return(FALSE)
+    x <- x[[f]]
+  }
+
+  TRUE
+}
+
+b_get <- function(brand, ...) {
+  if (!b_has(brand, ...)) {
+    return(NULL)
+  }
+
+  brand[[c(...)]]
+}
+
+b_has_string <- function(brand, ...) {
+  if (!b_has(brand, ...)) return(FALSE)
+  rlang::is_string(brand[[c(...)]])
+}
+
+b_has_list <- function(brand, ...) {
+  if (!b_has(brand, ...)) return(FALSE)
+  rlang::is_list(brand[[c(...)]])
+}
+
+# Find _brand.yml --------------------------------------------------------------
+
+find_project_brand_yml <- function(path = NULL) {
+  path <- path %||% getwd()
+  path <- normalizePath(path, mustWork = FALSE)
+
+  ext <- path_ext(path)
+  if (ext %in% c("yml", "yaml")) {
+    return(path)
+  } else if (nzchar(ext)) {
+    path <- dirname(path)
+  }
```

Comment:
```suggestion
  }
  
  if (nzchar(ext)) {
    path <- dirname(path)
  }
```

Reply:
lol I remember keeping the else because you tend to like the explicit branches. Happy to pull out the early return :)